tests/common/plugins/sanity_check/__init__.py

```diff
@@ -99,27 +99,41 @@ def sanity_check(localhost, duthost, request, fanouthosts, tbinfo):
                 json.dumps(check_results, indent=4))
-    if any([result["failed"] for result in check_results]):
-        if not allow_recover:
-            pytest.fail("Pre-test sanity check failed, allow_recover=False {}".format(check_results))
-            return
 
-        logger.info("Pre-test sanity check failed, try to recover, recover_method=%s" % recover_method)
-        recover(duthost, localhost, fanouthosts, check_results, recover_method)
-        logger.info("Run sanity check again after recovery")
-        new_check_results = do_checks(duthost, check_items)
-        logger.info("!!!!!!!!!!!!!!!! Pre-test sanity check after recovery results: !!!!!!!!!!!!!!!!\n%s" % \
-                    json.dumps(new_check_results, indent=4))
-        if any([result["failed"] for result in new_check_results]):
-            failed_items = json.dumps([result for result in new_check_results if result["failed"]], indent=4)
-            logger.error("Failed check items:\n{}".format(failed_items))
-            pytest.fail("Pre-test sanity check failed again after recovered by '{}' with failed items:\n{}".format(recover_method, failed_items))
+    pre_sanity_failed = False
+    for a_dutname, a_dut_results in check_results.items():
+        if any([result["failed"] for result in a_dut_results]):
+            pre_sanity_failed = True
+            if not allow_recover:
+                failed_items = json.dumps([result for result in a_dut_results if result["failed"]], indent=4)
+                logger.error("On {}, failed pre-sanity check items with allow_recover=False:\n{}".format(a_dutname, failed_items))
+            else:
+                logger.info("Pre-test sanity check failed on %s, try to recover, recover_method=%s" % (a_dutname, recover_method))
+                recover(duthosts[a_dutname], localhost, fanouthosts, a_dut_results, recover_method)
+
+        if not allow_recover and pre_sanity_failed:
+            pytest.fail("Pre-test sanity check failed on DUTs, allow_recover=False:{}".format(check_results))
             return
 
+    logger.info("Run sanity check again after recovery")
+    new_check_results = do_checks(duthosts, check_items)
+    logger.info("!!!!!!!!!!!!!!!! Pre-test sanity check after recovery results: !!!!!!!!!!!!!!!!\n%s" % \
+                json.dumps(new_check_results, indent=4))
+    pre_sanity_failed_after_recover = False
+    for a_dutname, a_dut_new_results in new_check_results.items():
+        if any([result["failed"] for result in a_dut_new_results]):
+            pre_sanity_failed_after_recover = True
+            failed_items = json.dumps([result for result in a_dut_new_results if result["failed"]], indent=4)
+            logger.error("On {}, failed check items after recover:\n{}".format(a_dutname, failed_items))
+
+    if pre_sanity_failed_after_recover:
+        pytest.fail("Pre-test sanity check failed on DUTs after recover:\n{}".format(new_check_results))
+        return
```

Comment:
Please change this ifblock to pytest_assert

Reply:
The idea was to log the failures of all the DUT's and then if any failed, then fail the checks. Is pytest_assert going to have similar affect.

Reply:
pytest_assert is a wrapper for if (! condition) pytest.fail.